lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,662 @@
+  {
+	  for(HornClauseDB::RuleVector::iterator it = db.getRules().begin(); it != db.getRules().end(); ++it)
+	  {
+			HornRule r = *it;
+			Expr rhead_app = r.head();
+			Expr rhead_cand_app = fAppToCandApp(rhead_app);
+			LOG("candidates", errs() << "HEAD: " << *rhead_app << "\n";);
+			LOG("candidates", errs() << "CAND: " << *rhead_cand_app << "\n";);
+			if(!isOpX<TRUE>(rhead_cand_app))
+			{
+				LOG("candidates", errs() << "ADD CONSTRAINT\n";);
+				db.addConstraint(rhead_app, rhead_cand_app);
+			}
+	  }
+  }
+
+  /*
+   * Use template to convert a predicate to candidate
+   */
+  Expr Houdini::relToCand(Expr fdecl)
+  {
+	ExprVector bvars;
+	ExprVector bins;   // a vector of LT expressions
+	Expr cand = NULL;
+
+	int bvar_count = 0;
+	unsigned i = 0;
+	for (i=0; i < bind::domainSz(fdecl); i++)
+	{
+		// if its type is INT
+		if (isOpX<INT_TY>(bind::domainTy(fdecl, i)))
+		{
+			// what is efac?
+			Expr bvar = bind::bvar (i, mk<INT_TY>(bind::domainTy(fdecl, i)->efac())); //the id of bvar is the same as related arg index
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		/*Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());*/
+		cand = mk<LT>(bvars[0], bvars[1]);
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+
+	return cand;
+  }
+
+  /*
+   * Build the map from predicates to candidates
+   */
+  void Houdini::guessCandidate(HornClauseDB &db)
+  {
+	  for(Expr rel : db.getRelations())
+	  {
+		  if(bind::isFdecl(rel))
+		  {
+			  Expr cand = relToCand(rel);
+			  currentCandidates.insert(std::pair<Expr, Expr>(rel, cand));
+		  }
+	  }
+  }
+
+  /*
+   * Main loop of Houdini algorithm
+   */
+  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm, HornClauseDBWto &db_wto, int config)
+  {
+	  guessCandidate(db);
+
+	  LOG("houdini", errs() << "CAND MAP:\n";);
+	  LOG("houdini", errs() << "MAP SIZE: " << currentCandidates.size() << "\n";);
+	  for(std::map<Expr, Expr>::iterator it = currentCandidates.begin(); it!= currentCandidates.end(); ++it)
+	  {
+		LOG("houdini", errs() << "PRED: " << *(it->first) << "\n";);
+		LOG("houdini", errs() << "CAND: " << *(it->second) << "\n";);
+	  }
+
+	  std::list<HornRule> workList;
+	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
+	  workList.reverse();
+
+	  if (config == EACH_RULE_A_SOLVER)
```

Comment:
why not use sub-classing or some other object-oriented or template-based solution?